### ISSUE TITLE
Support for security configuration in Servlet via Spring Security

### ISSUE THREAD

**Issue by dsyer** [spring-projects-experimental/spring-grpc#3]

**Comment by CyberZujo** [comment]
Hi @dsyer 
What would be the approach on this one? 
Should it be able to integrate spring-security through autoconfig and to have an interceptor for incoming requests on grpc server?

**Comment by dsyer** [comment]
If you wait for the interceptor support it will probably make it easier. I think the general idea is to create an interceptor that calls out to Spring Security. You can see the existing implementations at https://github.com/LogNet/grpc-spring-boot-starter and https://github.com/grpc-ecosystem/grpc-spring.

**Comment by onobc** [comment]
Until the interceptor support is officially added (should be in the next couple days), you can register a `ServerBuilderCustomizer` that adds the interceptor to the builder. Then when the interceptors are added we can go back and replace the customizer w/ an interceptor directly.

**Comment by CyberZujo** [comment]
Got a small example. 

```
        @Bean
	public SecurityInterceptor securityInterceptor() {
		return new SecurityInterceptor();
	}

	@Bean
	public ServerBuilderCustomizer<?> securityInterceptorCustomizer(SecurityInterceptor securityInterceptor) {
		return new SecurityInterceptorCustomizer<>(securityInterceptor);
	}
	
	
     public class SecurityInterceptor implements ServerInterceptor {

      @Override
      public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call, Metadata headers, 
      ServerCallHandler<ReqT, RespT> next) {
        Context context = Context.current();
        System.out.println("Intercepted GRPC call");
        return Contexts.interceptCall(context, call, headers, next);
    }
}
	
```
	
<img width="865" alt="image" src="https://github.com/user-attachments/assets/a2b4a8f6-e496-4b71-bfa1-481eda229ba5">

@onobc What do you think? Good direction?

**Comment by onobc** [comment]
Hi @CyberZujo , the above does look like an empty interceptor impl that could be used to hook into security. Although, you can also use a `@GlobalServerInterceptor` annotation instead of the customizer.

The delicate work here will be figuring out exactly what/how we want to hook into Spring Security. From the examples given by @dsyer above, we can see both the gRPC ecosystem starter ([here](https://github.com/grpc-ecosystem/grpc-spring/tree/master/grpc-server-spring-boot-starter/src/main/java/net/devh/boot/grpc/server/security)) and the Lognet starter ([here](https://github.com/LogNet/grpc-spring-boot-starter/tree/master/grpc-spring-boot-starter/src/main/java/org/lognet/springboot/grpc/security)) both have quite a bit of security components. My suggestion would be to first digest each of these and summarize what they do/not cover and from that we can decide which direction to go. Another option would be to just move the gRPC ecosystem security components directly in. 

@dsyer wdyt?

**Comment by onobc** [comment]
>  I'm also slightly wondering if we should implement something with the provided token-based authentication from grpc-java and leave out Spring Security completely unless someone asks for it.

If the token-based auth is sufficient then going w/ a single simpler thing for v1 would be a great idea IMO.

**Comment by CyberZujo** [comment]
@dsyer Yes, found it at: https://github.com/grpc-ecosystem/grpc-spring/blob/master/grpc-server-spring-boot-starter/src/main/java/net/devh/boot/grpc/server/security/interceptors/DefaultAuthenticatingServerInterceptor.java#L61

**Comment by dsyer** [comment]
That wasn't what I meant. Try starting here: https://github.com/grpc/grpc-java/blob/master/auth/

**Comment by CyberZujo** [comment]
Got it, sorry. Will proceed with that approach.

**Comment by dsyer** [comment]
I also noticed that JWT is only supported through a sample (https://github.com/grpc/grpc-java/blob/master/examples/example-jwt-auth). Maybe a Spring Security interceptor wouldn't be such a bad idea? The JWT sample is probably a good starting point anyway.

**Comment by benallard** [comment]
Ideally, one would _just_ add the `spring-boot-starter-oauth2-resource-server` starter, configure the `issuer`, and could start using the SecurityContextHolder. That is similar to the approach taken by the grpc-ecosystem starter.

**Comment by dsyer** [comment]
That could work. The dependency on spring-web and/or the servlet API might mess things up.

**Comment by benallard** [comment]
That's correct, the resource-server starter 's configuration classes are dependent on the Servlet web-application type. I usually copy the content  of `org.springframework.boot.autoconfigure.security.oauth2.resource.servlet.OAuth2ResourceServerJwtConfiguration` in my project to have it working.

**Comment by dsyer** [comment]
I started work on a security interceptor: https://github.com/dsyer/spring-grpc/tree/secure. It's quite tough going, but we probably don't need a lot of features to start with.

**Comment by dsyer** [comment]
Support for servlet apps added here: 6dc5d85. Native mTLS might be enough for the Netty-based servers for now.

**Comment by benallard** [comment]
Thanks @dsyer for your work on that. Am I wrong to believe there is no Netty support for reading-out Bearer Access-Tokens yet ? (Only Servlet via the standard servlet filters ?)

**Comment by dsyer** [comment]
Correct. Seems like a good solution to me, but feedback welcome. We can still maybe add some stuff for native servers with #93 (my https://github.com/dsyer/spring-grpc/tree/secure is still there). Google also has a sample in the `grpc-java` project as well (not using Spring Security obviously).
